Source/Applications/ApplicationArtifactIdentifier.cs

```diff
@@ -63,17 +63,23 @@ public bool Equals(IApplicationArtifactIdentifier other)
         {
             if (!Location.Equals(other.Location)) return false;
 
-            if (((string) Application.Name) != ((string) other.Application.Name)) return false;
-            if (((string) Artifact.Name) != ((string) other.Artifact.Name)) return false;
+            if (Application.Name.Value != other.Application.Name.Value) return false;
             if (Area.Value != other.Area.Value ) return false;
 
+            if (Artifact.Name.Value != other.Artifact.Name.Value) return false;
+            if (Artifact.Type.Identifier != other.Artifact.Type.Identifier) return false;
+            if (Artifact.Generation.Value != other.Artifact.Generation.Value) return false;
+
             return true;
         }
 
         /// <inheritdoc/>
         public override int GetHashCode()
         {
             var hashCode = Application.Name.GetHashCode();
+            // TODO: Shouldnt hashCode += Artifact.GetHashCode() instead?
```

Comment:
Absolutely correct. Much better solution - and putting the responsibility where it should be!
You could in fact do that with everyone of these elements - that makes much more sense. Rather than the line above even; `Application.Name.GetHashCode()` you could do `Application.GetHashCode()` - then those can live their own lives without this having to know about its internals. 

Same for the `Equals()` method - just rely on the objects themselves to guarantee correct equality comparison.